### PR TITLE
feat: lpg introspection route rules - close #89

### LINKED ISSUE #89: add LPG Gateway introspection to add LPG specific routes in route_rules
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add LPG gateway introspection to add LPG-specific routes to other gateways route_rules by adding a predefined value `lpg`. This is the same approach used for  other gateways with predefined values such as `drg`, `internet_gateway`, etc..

### New or Affected Resource(s)

oci_core_route_table * in [](vcn_gateway.tf)
oci_core_local_peering_gateway

### Potential Terraform Configuration

```hcl
  local_peering_gateways  = {lpg1_name = { route_table_id = "", peer_id = "possible_id_of_peered" }}
  nat_gateway_route_rules = [
    {
      destination       = "172.18.0.0/16",
      destination_type  = "CIDR_BLOCK",
      network_entity_id = "nat_gateway",
      description       = "already works, uses the id of the NAT GW created by the OKE module"
    },
    {
      destination       = "172.19.0.0/16",
      destination_type  = "CIDR_BLOCK",
      network_entity_id = "LPG@lpg1_name",
      description       = "this doesn't work right now, the network_entity_id would be the id of the LPG created above"
    }
  ]
```

### References

Initially raised here: oracle-terraform-modules/terraform-oci-oke#514

### PR DESCRIPTION
**Pull request by thpham**
# Proposed change

This PR resolve issue reported in the following tickets:
- #89 
- https://github.com/oracle-terraform-modules/terraform-oci-oke/issues/514

### Manual testing

If no automated testing is run, please ensure that at least the three steps below are passing without any error.

- [x] Running `terraform apply` on each example provided with this module provisions the intended resource(s) without any errors.
- [ ] Modifying module's *Input Variables* after initial provisioning behaves as intended, i.e: any updateable properties are ameneded without recreation of the resource(s).
- [x] Running `terraform destroy` on each example provided with this module destroys all the resources created by this module and only the resources created by this module.

## Checklist before submitting your PR

- [x] My code follows [the style guidelines of this project](../tree/main/docs/codingconventions.adoc)
- [ ] these changes provision new resources:
  - [ ] I have updated the README introduction section (README.adoc)
  - [ ] I have updated the README introduction section (README.md)
- [ ] these changes adds any new variables:
  - [ ] I have updated [docs/terraformoptions.adoc](../tree/main/docs/terraformoptions.adoc) to include each variable
- [ ] I have updated the changelog to include an entry for these changes
- [x] I have updated all provided examples, including each README file and all applicable code blocks
- [x] these changes generates no new warnings
- [ ] Any dependent changes have been merged and published in upstream modules

Note: *If you are not an Oracle employee, to contribute to an Oracle-sponsored open-source project, you need to sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/).*

### CODE REVIEW

File: examples/custom_route_rules/terraform.tfvars.example [hunk 1]
```diff
@@ -47,6 +47,10 @@ defined_tags = {
   "Operations.CostCenter" = "42"
 }
 
+local_peering_gateways = {
+  to_shared_svc = {}
```
Comment (hyder) [inline examples/custom_route_rules/terraform.tfvars.example:51]: Can you please also give a full blown configuration of what goes inside the ```to_shared_svc``` ?

**Comment by hyder** [comment]
This is like having a named variable for the LPG and using the named variable instead of just the ```lpg``` string or the LPG OCID. Am I correct in understanding that?

File: examples/custom_route_rules/terraform.tfvars.example (hunk 1 above)
Comment (thpham) [reply examples/custom_route_rules/terraform.tfvars.example:51]: in this situation for the simplicity of the example, we could imagine leaving it as is. This mean the module will create an LPG and will stay in acceptor state, until a peering request is done from another VCN for example.

Some documentation already exists that explain the various usage: do you want me to put a comment with the following link: https://github.com/oracle-terraform-modules/terraform-oci-vcn/blob/main/examples/hub-spoke/README.md#how-to-declare-one-or-many-lpg-on-the-vcn-module ?

**Comment by thpham** [comment]
> This is like having a named variable for the LPG and using the named variable instead of just the `lpg` string or the LPG OCID. Am I correct in understanding that?

This is a bit what you said, but as the input variable `local_peering_gateways` is a map that can create multiple named LPG, I think @snafuz proposed to use the prefix pattern `lpg@` in order to categorise the `route_rule` than need to be created and linked with the correct `network_entity_id` of the `oci_core_local_peering_gateway.lpg` for_each loop.

File: examples/custom_route_rules/terraform.tfvars.example (hunk 1 above)
Comment (hyder) [reply examples/custom_route_rules/terraform.tfvars.example:51]: Please give both in the terraform.tfvars.example and comment them. Something like:

```
local_peering_gateways = {
   # leave empty in acceptor state
   # to_shared_svc = {}
   
   # include parameter details below
   #to_lpg_{
      ...
      ...
   # }
}

Maybe we should also add these in the docs?

**Review by hyder — APPROVED** [review]

**Comment by hyder** [comment]
Thank you for your contribution

**Comment by thpham** [comment]
You're welcome ! May I ask you, what is your next release schedule for this module ?

**Comment by hyder** [comment]
Done: https://github.com/oracle-terraform-modules/terraform-oci-vcn/releases/tag/v3.6.0 :)

**Pull request merged**